analysis/pfe_methods/unicode_range_pfe_method.py

```diff
@@ -0,0 +1,129 @@
+"""Implementation of PFE that breaks a font into a set of language based subsets.
+
+A single font gets broken into one or more subsets (typically one subset per script).
+The client then chooses to download (via unicode range) the set of subsets which
+contain characters on in use on a given page. For a single session downloaded subsets
+are cached and re-used.
+
+Cut subsets are woff2 encoded to minimize their size.
+
+Subset definitions are provided in //analysis/pfe_methods/unicode_range_data.
+These definitions are based on what Google Fonts uses for their production
+font serving.
+"""
+
+import io
+import os
+
+from analysis import request_graph
+from analysis.pfe_methods.unicode_range_data import slicing_strategy_loader
+from fontTools import subset
+from woff2_py import woff2
+
+# Cache of which slicing strategy to use per font. Keyed by font name.
+FONT_SLICING_STRATEGY_CACHE = dict()
+
+# Cache of cut and woff2 encoded subset sizes. Keyed by:
+# <font name>:<slicing_strategy>:<subset index>
+SUBSET_SIZE_CACHE = dict()
+
+
+def name():
+  return "UnicodeRange"
+
+
+def start_session(font_directory, subset_sizer=None):
+  return UnicodeRangePfeSession(font_directory, subset_sizer)
+
+
+def slicing_strategy_for_font(font_id, font_bytes):
+  """Returns the slicing strategy that should be used to segment font_bytes."""
+  if font_id not in FONT_SLICING_STRATEGY_CACHE:
+    strategy_name = slicing_strategy_loader.slicing_strategy_for_font(
+        font_bytes)
+    FONT_SLICING_STRATEGY_CACHE[font_id] = strategy_name
+
+  strategy_name = FONT_SLICING_STRATEGY_CACHE[font_id]
```

Comment:
perhaps another `@lru_cache` opportunity?

Reply:
This method isn't a good fit since it takes font_bytes as an argument, but only cache_key should be used as the key.